addons/benchmark/benchmark.cc

```diff
@@ -194,87 +268,87 @@ double run_calc_minv_times_tau_benchmark (Model *model, int sample_count) {
   }
 
-  double duration = timer_stop (&tinfo);
+  report_constraints_run(*model, sample_data, "ForwardDynamicsConstraintsDirect");
 
-  return duration;
+  return sample_data.durations.sum();
 }
 
 double run_contacts_lagrangian_sparse_benchmark (Model *model, ConstraintSet *constraint_set, int sample_count) {
   SampleData sample_data;
   sample_data.fillRandom(model->dof_count, sample_count);
 
   TimerInfo tinfo;
-  timer_start (&tinfo);
 
   for (int i = 0; i < sample_count; i++) {
-    ForwardDynamicsConstraintsRangeSpaceSparse (*model, sample_data.q[i], sample_data.qdot[i], sample_data.tau[i], *constraint_set, sample_data.qddot[i]); 
+    timer_start (&tinfo);
+    ForwardDynamicsConstraintsRangeSpaceSparse (*model, sample_data.q[i], sample_data.qdot[i], sample_data.tau[i], *constraint_set, sample_data.qddot[i]);
+    sample_data.durations[i] = timer_stop (&tinfo);
   }
 
-  double duration = timer_stop (&tinfo);
+  report_constraints_run(*model, sample_data, "ForwardDynamicsConstraintsRangeSpaceSparse");
 
-  return duration;
+  return sample_data.durations.sum();
 }
 
 double run_contacts_null_space (Model *model, ConstraintSet *constraint_set, int sample_count) {
   SampleData sample_data;
   sample_data.fillRandom(model->dof_count, sample_count);
 
   TimerInfo tinfo;
-  timer_start (&tinfo);
 
   for (int i = 0; i < sample_count; i++) {
-    ForwardDynamicsConstraintsNullSpace (*model, sample_data.q[i], sample_data.qdot[i], sample_data.tau[i], *constraint_set, sample_data.qddot[i]); 
+    timer_start (&tinfo);
+    ForwardDynamicsConstraintsNullSpace (*model, sample_data.q[i], sample_data.qdot[i], sample_data.tau[i], *constraint_set, sample_data.qddot[i]);
+    sample_data.durations[i] = timer_stop (&tinfo);
   }
 
-  double duration = timer_stop (&tinfo);
+  report_constraints_run(*model, sample_data, "ForwardDynamicsConstraintsNullSpace");
 
-  return duration;
+  return sample_data.durations.sum();
 }
 
 double run_contacts_kokkevis_benchmark (Model *model, ConstraintSet *constraint_set, int sample_count) {
   SampleData sample_data;
   sample_data.fillRandom(model->dof_count, sample_count);
 
   TimerInfo tinfo;
-  timer_start (&tinfo);
 
   for (int i = 0; i < sample_count; i++) {
-    ForwardDynamicsContactsKokkevis(*model, sample_data.q[i], sample_data.qdot[i], sample_data.tau[i], *constraint_set, sample_data.qddot[i]); 
+    timer_start (&tinfo);
+    ForwardDynamicsContactsKokkevis(*model, sample_data.q[i], sample_data.qdot[i], sample_data.tau[i], *constraint_set, sample_data.qddot[i]);
+    sample_data.durations[i] = timer_stop (&tinfo);
   }
 
-  double duration = timer_stop (&tinfo);
+  report_constraints_run(*model, sample_data, "ForwardDynamicsContactsKokkevis");
 
-  return duration;
+  return sample_data.durations.sum();
 }
 
-double contacts_benchmark (int sample_count, ContactsMethod contacts_method) {
+void contacts_benchmark (int sample_count, ContactsMethod contacts_method) {
```

Comment:
Don't like the mix of model and contact method though the code is fine and functional. Suggest to have a refactoring PR for a second round.

Suggestions:

- Consider providing a std::function with the respective method from a list of function that you want to benchmark. Can share sample code.
- Same for the models. Consider a list of models where each element is benchmarked.

Reply:
I agree. I created a separate issue for this: #10